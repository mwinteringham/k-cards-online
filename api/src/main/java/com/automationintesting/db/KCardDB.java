package com.automationintesting.db;

import org.h2.jdbcx.JdbcDataSource;
import org.springframework.stereotype.Component;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;

@Component
public class KCardDB {

    private Connection connection;

    private final String CREATE_DB = "CREATE table WORKSHOPS ( id int NOT NULL AUTO_INCREMENT, code varchar(6), workshopName varchar(255), primary key (id))";

    public KCardDB() throws SQLException {
        JdbcDataSource ds = new JdbcDataSource();
        ds.setURL("jdbc:h2:mem:kcard;MODE=MySQL");
        ds.setUser("user");
        ds.setPassword("password");
        connection = ds.getConnection();

        connection.prepareStatement(CREATE_DB).executeUpdate();
    }

    public Boolean addCode(String code, String workshopName) throws SQLException {
        InsertWorkshopSql insertWorkshopSql = new InsertWorkshopSql(connection, code, workshopName);
        PreparedStatement createPs = insertWorkshopSql.getPreparedStatement();

        return createPs.executeUpdate() > 0;
    }
}
