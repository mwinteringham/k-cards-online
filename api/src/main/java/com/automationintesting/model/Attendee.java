package com.automationintesting.model;

import com.automationintesting.service.CodeGenerator;
import com.fasterxml.jackson.annotation.JsonProperty;

public class Attendee {

    @JsonProperty
    private String name;
    @JsonProperty
    private String code;

    public Attendee() {}

    public Attendee(String name, String code) {
        this.name = name;
        this.code = code;
    }

    public Attendee(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getCode() {
        return code;
    }

    public void setCode(String code) {
        this.code = code;
    }

    @Override
    public String toString() {
        return "Attendee{" +
                "name='" + name + '\'' +
                ", code='" + code +
                '}';
    }
}
