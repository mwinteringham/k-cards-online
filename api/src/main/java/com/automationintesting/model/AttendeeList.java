package com.automationintesting.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;

public class AttendeeList {

    @JsonProperty
    private List<Attendee> attendees;

    public AttendeeList() {
    }

    public AttendeeList(List<Attendee> attendees) {
        this.attendees = attendees;
    }

    public List<Attendee> getAttendees() {
        return attendees;
    }

    public void setAttendees(List<Attendee> attendees) {
        this.attendees = attendees;
    }

    @Override
    public String toString() {
        return "AttendeeList{" +
                "attendees=" + attendees +
                '}';
    }
}
