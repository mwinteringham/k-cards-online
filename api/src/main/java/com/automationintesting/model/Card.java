package com.automationintesting.model;

import com.automationintesting.service.CodeGenerator;
import com.fasterxml.jackson.annotation.JsonProperty;

public class Card {

    CodeGenerator codeGenerator = new CodeGenerator();

    @JsonProperty
    private String attendeeCode;
    @JsonProperty
    private String cardType;
    @JsonProperty
    private String cardCode;

    public Card() {
    }

    public Card(String attendeeCode, String cardType, String cardCode) {
        this.attendeeCode = attendeeCode;
        this.cardType = cardType;
        this.cardCode = cardCode;
    }

    public Card(String attendeeCode, String cardType) {
        this.attendeeCode = attendeeCode;
        this.cardType = cardType;
        this.cardCode = codeGenerator.createCode();
    }

    public String getCardType() {
        return cardType;
    }

    public void setCardType(String cardType) {
        this.cardType = cardType;
    }

    public String getAttendeeCode() {
        return attendeeCode;
    }

    public void setAttendeeCode(String attendeeCode) {
        this.attendeeCode = attendeeCode;
    }

    public String getCardCode() {
        return cardCode;
    }

    @Override
    public String toString() {
        return "Card{" +
                "attendeeCode='" + attendeeCode + '\'' +
                ", cardType='" + cardType + '\'' +
                ", cardCode='" + cardCode + '\'' +
                '}';
    }
}
