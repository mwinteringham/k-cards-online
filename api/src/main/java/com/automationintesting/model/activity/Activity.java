package com.automationintesting.model.activity;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;

public class Activity {

    @JsonProperty
    private List<CardDetail> reds;
    @JsonProperty
    private List<ActivityThread> threads;

    public Activity() {
    }

    public Activity(List<CardDetail> reds, List<ActivityThread> threads) {
        this.reds = reds;
        this.threads = threads;
    }

    public List<CardDetail> getReds() {
        return reds;
    }

    public void setReds(List<CardDetail> reds) {
        this.reds = reds;
    }

    public List<ActivityThread> getThreads() {
        return threads;
    }

    public void setThreads(List<ActivityThread> threads) {
        this.threads = threads;
    }

    @Override
    public String toString() {
        return "Activity{" +
                "reds=" + reds +
                ", threads=" + threads +
                '}';
    }
}
