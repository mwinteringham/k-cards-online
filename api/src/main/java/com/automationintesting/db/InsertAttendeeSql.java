package com.automationintesting.db;

import com.automationintesting.model.Attendee;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;

public class InsertAttendeeSql {

    private final PreparedStatement preparedStatement;

    public InsertAttendeeSql(Connection connection, Attendee attendee, String workshop) throws SQLException {
        final String CREATE_ATTENDEE = "INSERT INTO ATTENDEES (attendeecode, name, workshopcode) VALUES (?, ?, ?)";

        preparedStatement = connection.prepareStatement(CREATE_ATTENDEE);
        preparedStatement.setString(1, attendee.getCode());
        preparedStatement.setString(2, attendee.getName());
        preparedStatement.setString(3, workshop);
    }

    public PreparedStatement getPreparedStatement() {
        return preparedStatement;
    }

}
