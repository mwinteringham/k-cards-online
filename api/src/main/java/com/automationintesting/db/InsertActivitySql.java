package com.automationintesting.db;

import com.automationintesting.model.Card;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;

public class InsertActivitySql {

    private final PreparedStatement preparedStatement;

    public InsertActivitySql(Connection connection, Card card, String workshopCode) throws SQLException {
        final String CREATE_ACTIVITY = "INSERT INTO ACTIVITIES (cardtype, name, workshopcode) VALUES (?, ?, ?)";

        preparedStatement = connection.prepareStatement(CREATE_ACTIVITY);
        preparedStatement.setString(1, card.getCardType());
        preparedStatement.setString(2, card.getName());
        preparedStatement.setString(3, workshopCode);
    }

    public PreparedStatement getPreparedStatement() {
        return preparedStatement;
    }
}
