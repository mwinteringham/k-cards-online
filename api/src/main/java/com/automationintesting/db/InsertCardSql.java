package com.automationintesting.db;

import com.automationintesting.model.Card;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;

public class InsertCardSql {

    private final PreparedStatement preparedStatement;

    public InsertCardSql(Connection connection, Card card, String workshopCode) throws SQLException {
        final String CREATE_ACTIVITY = "INSERT INTO CARDS (cardtype, attendeecode, name, workshopcode) VALUES (?, ?, ?, ?)";

        preparedStatement = connection.prepareStatement(CREATE_ACTIVITY);
        preparedStatement.setString(1, card.getCardType());
        preparedStatement.setString(2, card.getAttendeeCode());
        preparedStatement.setString(3, card.getName());
        preparedStatement.setString(4, workshopCode);
    }

    public PreparedStatement getPreparedStatement() {
        return preparedStatement;
    }
}
