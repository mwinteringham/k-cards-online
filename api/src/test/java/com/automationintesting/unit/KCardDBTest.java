package com.automationintesting.unit;

import com.automationintesting.db.KCardDB;
import com.automationintesting.model.Attendee;
import com.automationintesting.model.AttendeeList;
import com.automationintesting.model.Card;
import com.automationintesting.model.activity.Activity;
import org.approvaltests.Approvals;
import org.junit.BeforeClass;
import org.junit.Test;

import java.sql.SQLException;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsEqual.equalTo;

public class KCardDBTest {

    private static KCardDB kCardDB;

    @BeforeClass
    public static void setupDb() throws SQLException {
        kCardDB = new KCardDB();
    }

    @Test
    public void storeWorkshopInDb() throws SQLException {
        String workshopCode = "abcdef";
        String workshopName = "LEWT";

        Boolean storeResult = kCardDB.addWorkshop(workshopCode, workshopName);

        assertThat(storeResult, equalTo(true));
    }

    @Test
    public void storeAttendeeInDb() throws SQLException {
        Attendee attendee = new Attendee("Stacey Jones");

        Boolean storeResult = kCardDB.addAttendee(attendee, "abcdef");

        assertThat(storeResult, equalTo(true));
    }

    @Test
    public void returnAttendeesList() throws SQLException {
        Attendee attendee = new Attendee("Barry White");
        kCardDB.addAttendee(attendee, "xyzxyz");

        AttendeeList attendeeList = kCardDB.getAttendeesInWorkshop("xyzxyz");

        assertThat(attendeeList.getAttendees().size(), equalTo(1));
    }

    @Test
    public void storeCardActivity() throws SQLException {
        Attendee attendee = new Attendee("Amy Lee");
        Card card = new Card(attendee.getName(), attendee.getCode(), "red");

        Boolean storeResult = kCardDB.addCardActivity(card, "jsdjsd");

        assertThat(storeResult, equalTo(true));
    }

    @Test
    public void returnWorkshopActivity() throws SQLException {
        Attendee attendee1 = new Attendee("Amy Lee");
        Attendee attendee2 = new Attendee("Stuart Jones");
        Attendee attendee3 = new Attendee("Sam Jones");

        Card redCard = new Card(attendee1.getName(), attendee1.getCode(), "red");
        Card greenCard = new Card(attendee2.getName(), attendee2.getCode(), "green");
        Card yellowCard = new Card(attendee3.getName(), attendee3.getCode(), "yellow");

        kCardDB.addCardActivity(redCard, "BEWT");
        kCardDB.addCardActivity(greenCard, "BEWT");
        kCardDB.addCardActivity(yellowCard, "BEWT");

        Activity activity = kCardDB.getWorkshopActivity("BEWT");

        Approvals.verify(activity);
    }

    @Test
    public void checkWorkshopExists() throws SQLException {
        kCardDB.addWorkshop("abcdef", "FEWT");

        Boolean workshopExists = kCardDB.doesWorkshopExist("abcdef");

        assertThat(workshopExists, equalTo(true));
    }

    @Test
    public void checkWorkshopDoesNotExist() throws SQLException {
        Boolean workshopExists = kCardDB.doesWorkshopExist("nowtef");

        assertThat(workshopExists, equalTo(false));
    }

    @Test
    public void removeAttendeeFromWorkshop() throws SQLException {
        Attendee attendee = new Attendee("James Salmon");
        kCardDB.addAttendee(attendee, "qwerty");

        Boolean attendeeRemoved = kCardDB.removeAttendee(attendee.getCode(),"qwerty");

        assertThat(attendeeRemoved, equalTo(true));
    }

    @Test
    public void removeAttendeesCards() throws SQLException {
        Attendee attendee = new Attendee("Boz Badger");

        Card redCard = new Card("Boz Badger", attendee.getCode(), "red");
        Card greenCard = new Card("Boz Badger", attendee.getCode(), "green");
        Card yellowCard = new Card("Boz Badger", attendee.getCode(), "yellow");

        kCardDB.addCardActivity(redCard, "poiuyt");
        kCardDB.addCardActivity(greenCard, "poiuyt");
        kCardDB.addCardActivity(yellowCard, "poiuyt");

        Boolean cardsRemoved = kCardDB.removeAttendeesCards(attendee.getCode(), "poiuyt");

        assertThat(cardsRemoved, equalTo(true));
    }
}
