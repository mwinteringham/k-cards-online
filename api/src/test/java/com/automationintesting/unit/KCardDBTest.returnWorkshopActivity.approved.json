{
  "reds": [
    {
      "code": "abcd",
      "name": "Amy Lee"
    }
  ],
  "threads": [
    {
      "cardDetail": {
        "code": "efgh",
        "name": "Stuart Jones"
      },
      "subthread": [
        {
          "code": "hijkl",
          "name": "Sam Jones"
        }
      ]
    }
  ]
}
